broadinstitute/single_cell_portal_core
# Dot plots, sorting, and pagination for differential expression table (SCP-4721, SCP-4900)

This lets users see DE metrics for _all_ genes right in SCP, sort them, and explore expression for many DE genes at once.

### Overview <span id="overview"></span>
Previously, only the 15 first rows from Scanpy's output were shown in the DE table.  This was useful as an MVP, but left a lot of typical table functionality unavailable, hampering its value for easily investigating differential gene expression patterns for a given group of cells.  There was also no quick way to investigate expression _across_ groups for many highly-ranked DE genes -- such many-gene, cross-group views for high-DE genes can help illuminate e.g. gene programs.

Now, users can explore more deeply through a handful of new features in the DE table:

* **Sort** by one or multiple columns.  By default, per requirements, rows are sorted by adjusted p-value then log2(FC).
* **Paginate** through DE metrics for all genes without leaving the app.
* **Dot plot** all genes on the current page of the DE table, to explore DE across groups of cells.

#### UI refinements <span id="ui-refinements"></span>
Additional UI refinements make exploration smoother.  The DE table is wider, which decreases horizontal crowding, and gives space for sort icons without sacrificing vertical real estate.  The table is now also more vertically responsive, heightening to accommodate available viewport tallness.  Up to 50 rows are visible.  When the viewport isn't tall enough to show all 50 rows, the user can scroll through the rows without losing sight of the data visualizations at left.  Modest rearrangements and negative space trimming help economize vertical real estate, to easily see and use more data and tools.

#### React Table
The engineering behind the table is also more robust.  The initial table was basic and hand-written.  To accommodate new and complex features, it made sense to port that to React Table.  An earlier version of this component is used elsewhere in the app; upgrading those instances seems worthwhile when doing work in those other areas.

### Video <span id="video"></span>
Here's how it looks.

#### Sort, scroll, and paginate DE table <span id="sort-scroll-and-paginate-de-table"></span>

https://user-images.githubusercontent.com/1334561/235714994-c8d0813c-4b8d-42b2-b2d9-1ebeb6039783.mov

#### Dot plot from DE table <span id="dot-plot-from-de-table"></span>

https://user-images.githubusercontent.com/1334561/236236334-f09d9960-0051-4598-92a8-2c09b0c720a6.mp4

_Edited via Blender to remove ~45 seconds waiting for dot plot data_

### Test <span id="test"></span>
An automated test confirms behavior.  Optionally, if you want to manually verify, follow the steps in the video, i.e.:

1.  **If local**, after pulling, since we're only testing frontend changes, try the frontend service worker cache enabled for faster page loads: `VITE_FRONTEND_SERVICE_WORKER_CACHE="true" bin/vite dev`. Then go to a study with DE results, e.g. the "mouse brain" synthetic study or (more representatively) the "Human milk - differential expression" DE pilot.  **If on staging**, go to the latter under [SCP303](https://singlecell-staging.broadinstitute.org/single_cell/study/SCP303/human-milk-differential-expression#study-visualize).
2. Click the blue "Differential expression" button, go to DE table
3. Confirm the DE table has adjusted p-value column in ascending order (smaller / more negative numbers closer to top) and the log2(fold change) column in descending order.
4. Click the sort icon in any column, confirm table updates accordingly
5. Click pagination icons at bottom of table, confirm table updates accordingly
6. Click dot plot icon, the first in the row of icons at top right of DE table
7. Confirm dot plot renders with genes on current page of DE table

### Further detail <span id="further-detail"></span>
See [SCP demo video](https://drive.google.com/file/d/12rqw54zd6yjhcdd7fjrFg92Mb4VIWL2C/view?t=31) for more context and discussion.  A follow-up to show the scrollbar in the DE table whenever rows exceed the viewing area is tracked in SCP-5119.  This satisfies SCP-4721 and SCP-4900.
